sympy/utilities/codegen.py

```diff
@@ -859,8 +913,47 @@ def _declare_globals(self, routine):
         return []
 
     def _declare_locals(self, routine):
-        # loop variables are declared in loop statement
-        return []
+
+        # Compose a list of symbols to be dereferenced in the function
+        # body. These are the arguments that were passed by a reference
+        # pointer, excluding arrays.
+        dereference = []
+        for arg in routine.arguments:
+            if isinstance(arg, ResultBase) and not arg.dimensions:
+                dereference.append(arg.name)
+        
+        code_lines = []
+        for result in routine.local_vars:
+
+            # local variables that are simple symbols such as those used as indices into
+            # for loops are defined declared elsewhere.
+            if not isinstance(result, Result):
+                continue
+
+            if result.name != result.result_var:
+                raise CodeGen("Result variable and name should match: {}".format(result))
+            assign_to = result.name
+            t = result.get_datatype('c')
+            if isinstance(result.expr, (MatrixBase, MatrixExpr)):
+                dims = result.expr.shape
+                if dims[1] != 1:
+                    raise CodeGenError("Only column vectors are supported in local variabels. Local result {} has dimensions {}".format(result, dims))
```

Comment:
tested. note that this should never happen using CSE as that only generates a list of assignments. it would be easy to declare a variable here as the right side, but the code below generates c code that unwraps that into a linear array. it seems risky to trust the memory layout when generating a two-dimensional float var[N1][N2] but would probably fly.

```
In [1]: from sympy import *
f
In [2]: from sympy.utilities.codegen import *

In [3]: a = MatrixSymbol('a', 2, 2)

In [4]: b = MatrixSymbol('b', 2, 2)

In [5]: r = Routine('a', [OutputArgument(a, a, b)], [], [Result(Matrix([[0, 1], [2, 3]]),b)], set())

In [6]: generator = CCodeGen(cse=True)

In [7]: generator.write([r], "output")[0][1]
---------------------------------------------------------------------------
CodeGenError                              Traceback (most recent call last)
<ipython-input-7-271abd7d1525> in <module>()
----> 1 generator.write([r], "output")[0][1]

~/python/sympy/sympy/utilities/codegen.py in write(self, routines, prefix, to_files, header, empty)
    760                 filename = "%s.%s" % (prefix, dump_fn.extension)
    761                 contents = StringIO()
--> 762                 dump_fn(self, routines, contents, prefix, header, empty)
    763                 result.append((filename, contents.getvalue()))
    764             return result

~/python/sympy/sympy/utilities/codegen.py in dump_c(self, routines, f, prefix, header, empty)
   1001 
   1002     def dump_c(self, routines, f, prefix, header=True, empty=True):
-> 1003         self.dump_code(routines, f, prefix, header, empty)
   1004     dump_c.extension = code_extension
   1005     dump_c.__doc__ = CodeGen.dump_code.__doc__

~/python/sympy/sympy/utilities/codegen.py in dump_code(self, routines, f, prefix, header, empty)
    801             code_lines.extend(self._declare_arguments(routine))
    802             code_lines.extend(self._declare_globals(routine))
--> 803             code_lines.extend(self._declare_locals(routine))
    804             if empty:
    805                 code_lines.append("\n")

~/python/sympy/sympy/utilities/codegen.py in _declare_locals(self, routine)
    938                 dims = result.expr.shape
    939                 if dims[1] != 1:
--> 940                     raise CodeGenError("Only column vectors are supported in local variabels. Local result {} has dimensions {}".format(result, dims))
    941                 code_lines.append("{0} {1}[{2}];\n".format(t, str(assign_to), dims[0]))
    942                 prefix = ""

CodeGenError: Only column vectors are supported in local variabels. Local result Result(b, Matrix([
[0, 1],
[2, 3]]), b) has dimensions (2, 2)
```

Reply:
I see. Yeah, we shouldn't make any assumptions about memory layout for generated C code.